### DISCUSSION TITLE
Dataloader plugin resolve info

### DISCUSSION THREAD

**Discussion by ericjohney** [hayes/pothos#291] [Q&A] [answered]
I was wondering if it makes sense or is possible to add graphql resolve info as like a 3rd parameter on the dataloader load function. Something like this...

```ts
const User = builder.loadableObject('User', {
  load: (ids: string[], context: ContextType, info: GraphQLResolveInfo) => context.loadUsersById(ids, getSelectedFieldsFromInfo(info)),
  fields: (t) => ({
    id: t.exposeID('id', {}),
    username: t.exposeString('username', {}),
  }),
});
```

The use case here being avoiding over fetching from the underlying datasource by knowing which fields have been selected in this graphql query.

**Comment by hayes** [comment, marked as answer]
No, I omitting info is intentional.  The info object is specific to a field, while loading an entity through a data-loader is designed may be loading that entity for multiple fields concurrently, or the result may be re-used for another field in the future.  This means that any info object provided here would apply to once instance/use of the field, but would not necissarily apply to every field that loader/id combination may apply to.

Not sure if my explanation makes sense.

**Reply by ericjohney** [reply]
Thanks for the response. Yeah, that definitely makes sense. Stupid followup question then, with the plugin as it is, can i do something like this...
```ts
load: (loads: { id: string, fields: string[] }[]) => context.loadStuff(getIds(loads), getSelectedFields(loads))
```
passing returning an object from the parent resolver like `{id: 7, fields: ['name', 'email']}` or if i really need to do that, would it be best to just not use the dataloader plugin like this

**Reply by hayes** [reply]
I don't think that will work.  The type of the ids should be restricted to string | int | bigint.  The plugin currently works based in the assumption that primitives (number, stirng, bigint) are keys, while objects are resolved data.  This allows you to return either an object, or an id from a parent resolver, and ensure it always gets resolved to the data using the loader when using a loadableObject.

Might be able to add this as a feature, but this falls a bit outside the use cases the plugin was designed for. The plugin assumes an id corresponds to a specific object.  What you are describing is more of a batch loading mechanism without the re-use portion.  Not sure if this makes more sense as something this plugin should support, or as its own plugin.

**Comment by 2coo** [comment]
@ericjohney, How did you tackle this?

```typescript
reservoirCollection: t.loadable({
      type: ReservoirCollection,
      resolve: (parent) => {
        return parent.address;
      },
      load: async (addresses: string[]) => {
      // How to get chainId which is argument of the grand parent query?
        const data = await reservoirClient.getNftsCollectionDetails(chainId, {
          contract: addresses,
        });
        return data.collections;
      },
    }),
```

Is there any existing feature that I missed? @hayes

**Reply by hayes** [reply]
This gets into tricky territory.  GraphQL makes this hard, and pothos doesn't have anything built in to handle it.   You don't want to do this in load, because load could (potentially) aggregate things from different parts of the query with a different grandparent.  You would likely want to return something like `{chain}{address}` from the resolve method, then parse out and group the addresses in load.  The easiest way to get an argument from a parent field is to pass it down as part of the resolved value.  You can technically extract it from the `info` object, but doing so reliably is complex, and I don't know of a good library to handle all the complexity cleanly.

**Reply by 2coo** [reply]
Thanks for your response!

Decided to go with `{chain_id}:{address}` from the resolver method until someone comes up with a better solution.

```typescript
reservoirCollection: t.loadable({
      type: ReservoirCollection,
      resolve: (parent, _args) => {
        // HACK: To pass the chain_id to the load function
        // https://github.com/hayes/pothos/discussions/291#discussioncomment-8403458
        return `${parent.chain_id}:${parent.address}`;
      },
      load: async (chainAndAddresses: string[]) => {
        const chainId = parseInt(chainAndAddresses[0].split(":")[0]);
        const addresses = chainAndAddresses.map((a) => a.split(":")[1]);
        const data = await reservoirClient.getNftsCollectionDetails(chainId, {
          contract: addresses,
        });
        return data.collections;
      },
    }),
```
